### ISSUE TITLE
Spain

### ISSUE THREAD

**Issue by sbma44** [openaddresses/openaddresses#771]
[Spain's cadastral data seems to be open by decree](http://www.catastro.minhap.es/ayuda/legislacion/ovc/resoluciondgc20110323_tfs.pdf). Unfortunately, in [its easiest-to-use forms it is gated](https://www.sedecatastro.gob.es/OVCFrames.aspx?TIPO=TIT&a=masiv) by client-side certs that are only available to Spanish citizens who are using eServices.

Luckily, as part of the INSPIRE process, Spain has released an ungated version [here](http://blog-idee.blogspot.com.es/2014/03/actualizacion-de-los-datos-de-descarga.html). Less luckily, it consists of 7,000+ GML files, where the street name is made available through an XLINK record that `ogr2ogr` does not seem to resolve correctly. For now I am working under the assumption that this data is licensed under the same terms as the non-INSPIRE dataset. Some custom scripting will be necessary to resolve the street names and make this data processable by OA.

**NOTE:** OSM `talk-es` users have pointed me to some concerns about the quality of the cadastral dataset -- specifically, house numbers getting paired with the wrong street names. I'm not yet sure how big of a deal this is. Any Spanish speakers who can give the following discussion a closer read, by all means do so (I'm working off of Google Translate). 

http://gis.19327.n5.nabble.com/Catastro-duda-nombres-de-calles-td5777261.html

**Comment by sbma44** [comment]
reopening: looks like Spain has an open data portal at http://www.cartociudad.es/portal/ which might allow us to get away from the cadastral source.

**Comment by EamonKeane** [comment]
@sbma44 yes can confirm that works. I used that page to put together a full dataset without authentication required. This is done by downloading the 50 provinces from the 'CartoCity' dataset in the search by listing section.

The data returned is like this:
https://www.dropbox.com/s/54s9xvwqcd4g1s9/CARTOCIUDAD_CALLEJERO_ALAVA.zip?dl=0

<img width="423" alt="screen shot 2017-07-13 at 15 36 36" src="https://user-images.githubusercontent.com/1946655/28171825-9980d772-67e1-11e7-9fea-8388dc06e243.png">

By combining the PORTAL_PK shapefiles, you can reproduce the Portales y PK layer of this WMS:
http://www.cartociudad.es/wms/CARTOCIUDAD/CARTOCIUDAD?Request=GetCapabilities

I manually downloaded the 50 files, but I believe you could automate this process by using an AJAX scraper. I might look into this if I have time.

**Comment by EamonKeane** [comment]
I made a layer with the postcode centroid (average of all lat/lon with same postcode) which is useful as a backstop when geocoding unclear addresses (for route optimisation in particular). Sharing it here in case it's of use.

https://gist.github.com/EamonKeane/702011de5af18d2612f813533d792bc8

<img width="891" alt="screen shot 2017-07-20 at 11 15 59" src="https://user-images.githubusercontent.com/1946655/28414260-c70f2f48-6d42-11e7-8e27-e95f9fe174a0.png">

**Comment by migurski** [comment]
This is great, @EamonKeane! Should we think about replacing our Spain data with what you're working on?

**Comment by EamonKeane** [comment]
@migurski I don't know that much about how the data is sourced in the current OpenAddresses Layer vs the 'CartoCity' dataset. It appears that the 'CartoCity' dataset is cleaner than current OpenAddresses, however it has less entries. 

There are 14.5m entries in the current OpenAddresses dataset, however of these there are 1.7m which just have 'S/N' (without number). These don't add anything meaningful, and many of them appear to be inaccurate.

There are 10.9m entries in the dataset I downloaded (50 provinces). A portion of these are km markings along major roads (not sure how many). The source for this layer is 'Dirección General del Catastro', which I think is the same sources as the current OpenAddress data. 

The pros of switching might be that is more maintainable and more easily updatable. Not sure what has happened for legitimate houses to be present in the OA dataset but not in the current dataset. Could it be a change of methodology in the meantime? 

I could provide the dataset if that is any use to you?

Red is entries in the new dataset. The blue markers are the current OpenAddresses.

<img width="652" alt="screen shot 2017-07-21 at 11 48 24" src="https://user-images.githubusercontent.com/1946655/28461401-57975808-6e0e-11e7-92a1-8ad8ff8c556b.png">
<img width="1243" alt="screen shot 2017-07-21 at 11 45 41" src="https://user-images.githubusercontent.com/1946655/28461410-5fd33924-6e0e-11e7-9584-40bc588026d2.png">

**Comment by nvkelso** [comment]
For the address in the proposed `#ff0000` new CartoCity dataset they are more "front door" (and higher quality) than the existing `#0000ff` OpenAddresses dataset. Nice!

But it looks like the existing `#0000ff` OpenAddresses dataset has _more_ addresses that we wouldn't want to drop. But it also has lots of trash `S/N` (without number?) points. What do we usually do with points without a number, do we drop them?

Should we **add** the new CartoCity dataset and **keep** the existing dataset? We do that in some places in the USA where coverage is provided by multiple sources (mix of locality, county, regional, country-wide)?

**Comment by sbma44** [comment]
@nvkelso is your work available anywhere? I'm curious to see what deduping would require...

**Comment by nvkelso** [comment]
FYI: openvenues deduping for reference: https://github.com/openvenues/lieu

**Comment by thatdatabaseguy** [comment]
@nvkelso indeed, the  address-only version can detect dupes with many variations in street names, etc. in Castilian Spanish, Catalan, Galician, and Basque. It could be made to prefer the addresses in the CartoCity data set, then the `S/N` non-dupes from OpenAddresses could simply be removed as part of post-processing.

The one thing I need to add to make the process work well on Spanish addresses is a street field normalization which strips thoroughfare types and directionals altogether such that "Calle Ruiz" will match with simply "Ruiz". This may be less of an issue in Spain but in Latin America the "Calle" is very often omitted. It appears that in the current OpenAddresses, all of the street names are prefixed with thoroughfare type codes like "CL", "CR", "AV", etc. which libpostal can normalize, and if that's the case for the CartoCity data as well, then lieu should work great as is.

**Comment by EamonKeane** [comment]
@sbma44 the raw data is here (5 GB compressed, contains multiple datasets such as roads and building outlines for each province in addition to the addresses  - the 'PORTAL_PK.shp' layer):
https://www.dropbox.com/s/vcqkhl4x22kfs7y/Spain_CartoCity.zip?dl=0

The algorithm to get the data is:

- Extract each sub-folder representing a province
- Concatenate each of the 'PORTAL_PK.shp' files
- Dump to a csv or to a DB

A rough script for doing so is here:
https://gist.github.com/EamonKeane/16ce83ad702125224a723e677a363b09

A postgis dump is here (3.3GB - schema could be polished up):
https://www.dropbox.com/s/nmfdq2z8p33t48z/carto_ciudad.sql?dl=0

Deduping I think will be tricky. There are no place names, only house numbers and streets. The existing dataset includes street modifiers like 'CL COSTANA' as @thatdatabaseguy mentions, while the new one does not have them, just e.g. 'COSTANA'.

An algorithm for deduping might be:
- Remove S/Ns
- If the same house number is present in both datasets within say 50m, use the lat/lon from the new dataset (this appears to have more accurate positioning). This would be very processing intensive  for two 10M point datasets. Would openvenues/lieu be able to do that?
- Remove the 1km land markings in the new CartoCity dataset. These seem to have a roadname pattern to them, like 'LU-211', 'N-540' in the street field. Their lat-lon also intersects the TRAMO_VIAL road network shapefile.
- For the residual probably 1m-2m points that are present in he existing dataset, they could either all be included or a judgement could be made as to their quality perhaps using nearby roads or satellite imagery to decide.

**Comment by thatdatabaseguy** [comment]
@EamonKeane yes, lieu should be able to handle that fine on the next build once I've made the change to match without thoroughfare types like "CL". We have [fairly comprehensive dictionaries for Spanish](https://github.com/openvenues/libpostal/blob/master/resources/dictionaries/es) in terms of the types of phrases that can be removed from the OA street names.

The only requirements are a lat/lon, street and house number, ignores place names. As far as precedence, lieu treats the inputs as one big concatenated list of GeoJSON features, and prefers whichever record came first as the canonical (everything else is a dupe of that record). So if the CartoCity file is passed in first in the arguments list, the CartoCity lat/lons would be preferred over OA, while still allowing for records that were in OA and not in CartoCity to be included.

All of the other steps could be done as pre/post-processing.

**Comment by sbma44** [comment]
@aarthykc and I have had a closer look at this dataset. In all of these the older allegedly-cadastral data is in red, new dataset in green:

![image](https://user-images.githubusercontent.com/31717/34274898-c5fa728a-e668-11e7-8159-19b1186dfc50.png)
![image](https://user-images.githubusercontent.com/31717/34274936-fbc5943a-e668-11e7-804f-778d027b5a3f.png)
![image](https://user-images.githubusercontent.com/31717/34275075-b01c81c8-e669-11e7-8c3b-3a9779693a68.png)
![image](https://user-images.githubusercontent.com/31717/34275223-4e0050e0-e66a-11e7-8946-013eb901efde.png)

I'm not finding a lot of circumstances where the new dataset looks better. It's missing street type (possibly due to the export script needing adjustment). More to the point though the actual geometry often looks tied to the road network. We had worried about the older dataset being cadastral/parcel-y but I'm seeing what looks like rooftop points in at least some areas.

We've also reconfirmed that the total points (14,531,225 vs 9724884) and covered areas favor the older data:

![image](https://user-images.githubusercontent.com/10141319/34259043-b1cbbe18-e686-11e7-90f6-1d6f7871586c.png)
![image](https://user-images.githubusercontent.com/10141319/34259055-bc10120c-e686-11e7-9d84-16dab32e97d7.png)

All of this has led us to push import efforts to the back burner.

**Comment by frodrigo** [comment]
I made an update to the Cadastral source scripts in this PR #6968. The cadastral source now contains 15,825,571 addresses. Basque Country and Navarre are missing.

The new CartoCuidad mainly contains addresses in towns, not in rural part.

![Capture d’écran_2023-10-27_21-24-09](https://github.com/openaddresses/openaddresses/assets/1785486/cc863ee2-f4fb-4689-958c-3dde9ca2a390)
(Canary Islands out of the screen, but included in the data)
